### PR TITLE
Implement bech32, assetAddress, platformAddress

### PR DESCRIPTION
**Pull request by posgnu**

### CODE REVIEW

File: codechain/crypto/bech32.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
```
Comment (foriequal0) [inline codechain/crypto/bech32.py:22]: I think you've borrowed code from here: https://github.com/sipa/bech32/blob/master/ref/python/segwit_addr.py
There's a copyright notice(I think it is a MIT license). It says:
```
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
```
You should follow it. If you modified the code, then here's how you should do it:
https://softwareengineering.stackexchange.com/questions/157968/how-to-manage-a-copyright-notice-in-an-open-source-project
Or like this template: https://github.com/gmalecha/coq-plugin-template/blob/master/LICENSE

Comment (posgnu) [reply codechain/crypto/bech32.py:22]: I added the copyright claim.

File: .flake8 [hunk 2]
```diff
@@ -1,8 +1,9 @@
 [flake8]
-ignore = E501,F403,F401
+ignore = E501,F403,F401, E203
```
Comment (foriequal0) [inline .flake8]: This is a whole project exception just for a single expression `[i : i + ...]` in this PR. It is against the PEP-8 guide. Do you have any reasons for this?
https://www.python.org/dev/peps/pep-0008/#pet-peeves

File: .flake8 [hunk 3]
```diff
@@ -1,8 +1,9 @@
...
     # E501: Line too long
     # F403: Unable to detect undefined names
     # F401: Imported but unused
-max-line-length = 80
+    # E203: whitespace before ':'
+max-line-length = 100
```
Comment (foriequal0) [inline .flake8]: There already are lines that exceeds `max-line-length`. Can we leave it as is for this PR and make another PR?

File: codechain/primitives/AssetAddress.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        self.address_type = address_type
+
+        if isinstance(payload, (H160, str)):
+            self.payload = H160(payload)
+        else:
+            n, m, pubkeys = payload
+            self.payload = (n, m, list(map(lambda p: H160(p), pubkeys)))
+        self.value = address
+
+    @classmethod
```
Comment (foriequal0) [inline codechain/primitives/AssetAddress.py]: Prefer `@staticmethod` since we do not likely inherit AssetAddress.

File: codechain/primitives/AssetAddress.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    def from_type_and_payload(
+        cls, network_type: int, payload: Union[H160, str, MultiSig], **kwargs
+    ):
+        if "network_id" in kwargs:
```
Comment (foriequal0) [inline codechain/primitives/AssetAddress.py]:
```
network_id = kwargs.get("network_id", "cc")
version = kwargs.get("version", 1)
```
https://docs.quantifiedcode.com/python-anti-patterns/correctness/not_using_get_to_return_a_default_value_from_a_dictionary.html

File: codechain/primitives/AssetAddress.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+        if address_type < 0x3:
+            payload = H160(byte[2:])
+            return cls(address_type, payload, address)
+        else:
+            n = byte[2]
+            m = byte[3]
+            payload = byte[4:]
+            pubkeys = []
+            for i in range(0, len(payload), 20):
```
Comment (foriequal0) [inline codechain/primitives/AssetAddress.py:81]: Should we check `len(payload)` is multiple of 20?

File: codechain/primitives/AssetAddress.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    def check_string(cls, address: str):
+        return None is not re.match(
+            "^.{2}a[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{42}$", address
+        )
+
+    @classmethod
+    def ensure(cls, address):
+        return address if isinstance(address, cls) else cls.from_string(address)
+
+    def to_string(self) -> str:
```
Comment (foriequal0) [inline codechain/primitives/AssetAddress.py]: We can use `def __str__(self)` for this since it doesn't require any formating parameters.

File: codechain/primitives/PlatformAddress.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+                f"Invalid public key for creating PlatformAddress: {public_key}"
+            )
+
+        return cls.from_account_id(
+            get_account_id_from_public(str(public_key)), **kwargs
+        )
+
+    @classmethod
+    def from_account_id(cls, account_id, **kwargs):
+        if "network_id" in kwargs:
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py]:
```
network_id = kwargs.get("network_id", "cc")
verison = kwargs.get("version", "1")
```

File: codechain/primitives/PlatformAddress.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            network_id = kwargs.get("network_id")
+        else:
+            network_id = "cc"
+        if "version" in kwargs:
+            version = kwargs.get("version")
+        else:
+            version = 1
+
+        if not H160.check(account_id):
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py:31]: `H160.check` returns `True` if the format is correct regardless of type. It can be `str` or `H160`, but it contradicts with `__init__` requires `account_id` to be `H160`. You might want to use `account_id = H160(account_id)` to force the type, or just accept `H160` type.

File: codechain/primitives/PlatformAddress.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            return address
+        elif isinstance(address, str):
+            return cls.from_string(address)
+        else:
+            raise ValueError(
+                f"Expected either PlatformAddress or string but found {address}"
+            )
+
+    @classmethod
+    def ensure_account(cls, address):
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py]: Just ```PlatformAccount.ensure(x).account_id``` would be sufficient

File: codechain/primitives/PlatformAddress.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    @classmethod
+    def check_string(cls, value: str):
+        return None is not re.match(
+            "^.{2}c[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{40}$", value
+        )
+
+    def to_string(self):
+        return self.value
+
+    def get_account_id(self):
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py]: Getter for a value type would not be necessary

File: codechain/primitives/PlatformAddress.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
+import re
+
+from ..crypto import bech32_decode
+from ..crypto import bech32_encode
+from ..crypto import blake160
+from .HexString import H160
+from .HexString import H512
+
+
+class PlatformAddress:
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py]: PlatformAddress should be a value type. Inherit `typing.NamedTuple`.

File: codechain/primitives/PlatformAddress.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        else:
+            return cls.from_string(address).get_account_id()
+
+    @classmethod
+    def check_string(cls, value: str):
+        return None is not re.match(
+            "^.{2}c[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{40}$", value
+        )
+
+    def to_string(self):
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py]: Prefer ```def __str__```

File: codechain/primitives/PlatformAddress.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        return self.value
+
+    def get_account_id(self):
+        return self.account_id
+
+
+def get_account_id_from_public(public_key: str):
+    if not isinstance(public_key, str):
+        raise ValueError(
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py]: It should raise `TypeError`. Also `blake160` can accept both `str` and `bytelike`. Change `public_key` type to `Union[str, H512, bytes]` or leave blake160 to handle type error for now. Also you've used this method in ```PlatformAddress.from_public``` which involves unnecessary converting of `bytelike` into `str`

File: codechain/primitives/AssetAddress.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+from .HexString import H160
+
+
+class MultiSig(NamedTuple):
+    n: int
+    m: int
+    pubkeys: List[H160]
+
+
+class AssetAddress:
```
Comment (foriequal0) [inline codechain/primitives/AssetAddress.py:18]: It is useful to have `__hash__` and `__eq__` method. It allows `AssetAddress` to be used as a dictionary key. It would've been better to make it a value type (which gives you `__hash__` and `__eq__` and other methods for free, but it would be quite difficult to express in Python for this type.
https://stackoverflow.com/questions/4901815/object-of-custom-type-as-dictionary-key

**Review by foriequal0 — CHANGES_REQUESTED** [review]
Also
1. Make another PR for `Add to_string and to_json tests for Hex string`
2. Change title `Implement bech32` to somewhat `Modify bech32 implementation to match with CodeChain address format`

File: .flake8 (hunk 2 above)
Comment (posgnu) [reply .flake8]: I removed it.

File: .flake8 (hunk 3 above)
Comment (posgnu) [reply .flake8]: Okay

File: codechain/primitives/AssetAddress.py (hunk 4 above)
Comment (GNUp) [reply codechain/primitives/AssetAddress.py]: Okay.

Comment (posgnu) [reply codechain/primitives/AssetAddress.py]: But this just said https://stackoverflow.com/questions/1950414/what-does-classmethod-do-in-this-code/1950927#1950927 one of the main uses of classmethod is to define alternative constructors.

File: codechain/primitives/AssetAddress.py (hunk 5 above)
Comment (posgnu) [reply codechain/primitives/AssetAddress.py]: Thanks.

File: codechain/primitives/AssetAddress.py (hunk 6 above)
Comment (posgnu) [reply codechain/primitives/AssetAddress.py:81]: I made a check routine.

File: codechain/primitives/AssetAddress.py (hunk 7 above)
Comment (posgnu) [reply codechain/primitives/AssetAddress.py]: Thanks

File: codechain/primitives/PlatformAddress.py (hunk 8 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: I changed it.

File: codechain/primitives/PlatformAddress.py (hunk 9 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py:31]: I changed function signature.
```
    def from_account_id(cls, account_id: H160, **kwargs):
```

File: codechain/primitives/PlatformAddress.py (hunk 10 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: Okay, I removed it.

File: codechain/primitives/PlatformAddress.py (hunk 11 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: I removed it.

File: codechain/primitives/PlatformAddress.py (hunk 12 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: what is the value type?

File: codechain/primitives/PlatformAddress.py (hunk 13 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: I changed it.

File: codechain/primitives/PlatformAddress.py (hunk 14 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: I removed unnecessary checking routines.

File: codechain/primitives/AssetAddress.py (hunk 15 above)
Comment (posgnu) [reply codechain/primitives/AssetAddress.py:18]: I added `__hash__` method.

File: codechain/primitives/AssetAddress.py (hunk 4 above)
Comment (foriequal0) [reply codechain/primitives/AssetAddress.py]: It might be useful when we let `AssetAddress` inheritable. It'll work like an alternative `__new__`, which gives you the `cls` parameter too for similar reasons. 
However, it doesn't seem to useful to inherit it. We don't know what invariants are applied to its subclasses. That's why many libraries make classes `final`, which make them non-inheritable. 
In this case, `@classmethod` only leaves unnecessary access to `cls`.

File: codechain/primitives/PlatformAddress.py (hunk 12 above)
Comment (foriequal0) [reply codechain/primitives/PlatformAddress.py]: A value type is a type that can be compared by values of fields, not by the reference.

```python
def foo():
	return 10000000 + 10000000
def bar():
	return 20000000
a = foo()
b = bar()
assert a is not b
# they don't have the same reference 
# a, b, they are different objects that are allocated in different heap locations.
# Python's implementation detail:
# Python's integer is also an object. However, 
# Python caches and shares small integer reference, 
# but not for the large integers. They allocate it every time. 
# see: https://stackoverflow.com/questions/15171695/whats-with-the-integer-cache-maintained-by-the-interpreter)
assert a == b
# but they are compared by its value. 
# If the value is same, `__eq__` returns `true`, and `__hash__` is the same.
```

```python
class ComplexType:
	def __init__(self, value):
		self.value = value
	def __add__(self, other):
		return Wrapper(self.value + other.value)
def foo():
	return ComplexType(20000000)
def bar():
	return ComplexType(20000000)
a = foo()
b = bar()
assert a is not b 
# They are different entities that are allocated in different locations.
assert a != b 
# they are compared by the reference, even its fields are identical values.
```
To make `ComplexType` be compared by value, then we should implement `__eq__` and `__hash__`. Python already did it to `list`(`[10000 + 10000] == [20000]`), `tuple`(`(10000 + 10000,) == (20000,)`), `dict`, `set`, ... . Every complex types shouldn't have to be a value type, but some types are useful to be a value type.

File: codechain/primitives/PlatformAddress.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+from ..crypto import bech32_decode
+from ..crypto import bech32_encode
+from ..crypto import blake160
+from .HexString import H160
+from .HexString import H512
+
+
+class PlatformAddress(typing.NamedTuple):
+    def __init__(self, account_id: H160, address: str):
```
Comment (foriequal0) [inline codechain/primitives/PlatformAddress.py]: `NamedTuple` works by declaring fields and its types at the class level and it prohibits to overwrite some fields (https://github.com/python/typing/blob/3.6.1/src/typing.py#L2052)
```
In [3]: class Foo(NamedTuple):
   ...:     def __init__(self, a):
   ...:         self.a = a
   ...:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-349168dd385e> in <module>
----> 1 class Foo(NamedTuple):
      2     def __init__(self, a):
      3         self.a = a
      4

/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/typing.py in __new__(cls, typename, bases, ns)
   1372         for key in ns:
   1373             if key in _prohibited:
-> 1374                 raise AttributeError("Cannot overwrite NamedTuple attribute " + key)
   1375             elif key not in _special and key not in nm_tpl._fields:
   1376                 setattr(nm_tpl, key, ns[key])

AttributeError: Cannot overwrite NamedTuple attribute __init__
```

You should do this instead:
```python
class PlatformAddress(typing.NamedTuple):
    account_id: H160,
    value: str,

PlatformAddress(account_id=H160(...), value="...")
```

How did this passed the test?

File: codechain/primitives/AssetAddress.py (hunk 4 above)
Comment (posgnu) [reply codechain/primitives/AssetAddress.py]: Okay.

File: codechain/primitives/PlatformAddress.py (hunk 12 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: Thanks! good information.

**Review by foriequal0 — CHANGES_REQUESTED** [review]

**Review by foriequal0 — CHANGES_REQUESTED** [review]
There are still `@classmethod` left. Can you change that too?

File: codechain/primitives/PlatformAddress.py (hunk 16 above)
Comment (posgnu) [reply codechain/primitives/PlatformAddress.py]: Sorry, I forgot running the test. I fixed it.

**Review by foriequal0 — APPROVED** [review]
